### ISSUE TITLE
[Feature Request] Make possible to extend features with custom plugins

### ISSUE THREAD

**Issue by rcasta74** [yegappan/lsp#148]
I have just seen that a commit was added to define custom notification handlers; it would be nice to have a more generic way to:
1) override any handler
2) add a new lsp request

Examples of usage:
1) override `ShowReferences` to use `fzf` plugin (this will give a smoother look is `fzf` is already used for other commands)
2) add `java/classFileContents` request for jdt-ls (when decompiler boundle is enable in jdt-ls, it can be used to get the decompiled source for a reference with `jdt://` URI)

**Comment by rcasta74** [comment]
> 1. I think that it could make sense to be able to provide functions for rendering the references

Yeah, I was thinking to something like this. The function `lspserver.ShowReferences` currently send the reqest to the server and then it processes the reply calling the function `symbol.ShowReferences`. If instead it calls a registered function and the possibility to override the registered function is given, we can customize the way how the reply is processed.
`ShowReferences` was only an example, it could be done for all lsp requests or for a subset of them that could make sense to be overriden.

> 2. How would this work in a general purpose language server client? Is it a custom code action?

No, `java/classFileContents` is not a code action but a custom request available only for jdt-ls: the client send the request to the server and the server reply with the text of decompiled class.
Anyway looking deeper in the code it seems that adding a new request is already possible: with the exported function `buffer.CurbufGetServer` I can get the lspserver for the current buffer and then I have access  to the function `rpc` to send the request. I will give it a try in the next days.

**Comment by yegappan** [comment]
It is good idea to add hooks to allow using fzf with the language server results.  One problem I see is that the fzf custom hook (fzf#run) allows jumping to a file but not to a specific location in the file.  Need to figure out a way to jump to a particular location in the selected file.

**Comment by yegappan** [comment]
I have been thinking about separating the UI layer from the language server interface.  This will allow exposing APIs directly to the language server interface.  For users who want the LSP functionality out of the box, the built-in UI will be sufficient.  The advanced users can directly use the language server API and customize the UI.

**Comment by sigzegv** [comment]
Is extension possible new or not yet ? didn't saw anything in documentation.

**Comment by jclsn** [comment]
At least the second point is now possible via https://github.com/yegappan/lsp/pull/640
